### ISSUE TITLE
Fix build with ANGLE-only Qt on Windows

### ISSUE THREAD

**Issue by annulen** [annulen/webkit#421]
`Windows 10 (msvc2015-x86_64) 	DeveloperBuild, Release, QtNamespace, QtLibInfix`

This is not configuration used in Qt SDK, but it is tested in Coin.

If I force -DUSE_OPENGL_ES_2=1 compilation fails with following error:

```
	C:\PROGRA~2\MICROS~1.0\VC\bin\amd64\cl.exe  @C:\Users\qt\AppData\Local\Temp\TemporaryOpenGLSetting.cpp.obj.3112.2018469.jom
TemporaryOpenGLSetting.cpp
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(60): error C3861: 'glIsEnabled': identifier not found
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(65): error C2039: 'glEnable': is not a member of '`global namespace''
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(65): error C3861: 'glEnable': identifier not found
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(67): error C2039: 'glDisable': is not a member of '`global namespace''
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(67): error C3861: 'glDisable': identifier not found
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(76): error C2039: 'glEnable': is not a member of '`global namespace''
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(76): error C3861: 'glEnable': identifier not found
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(78): error C2039: 'glDisable': is not a member of '`global namespace''
C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\TemporaryOpenGLSetting.cpp(78): error C3861: 'glDisable': identifier not found
``` 

If I don't it also fails:
```
 C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(113): error C2440: 'static_cast': cannot convert from 'WebCore::Extensions3D *' to 'WebCore::Extensions3DOpenGLES *'
 C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(113): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
 C:\Users\qt\work\qt\qtwebkit\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(114): error C2027: use of undefined type 'WebCore::Extensions3DOpenGLES'
...
```

Maybe it has something to do with namespaced Qt

**Comment by tomwilliz** [comment]
After applying the 2 patches to qtwebkit-tp5 I still get a compilation error:

```
D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGL.cpp(195): error C2065: 'GL_RGBA8': undeclared identifier [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGL.cpp(199): error C2065: 'GL_RGB8': undeclared identifier [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGL.cpp(211): error C2065: 'GL_DEPTH24_STENCIL8_EXT': undeclared identifier [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]

... etc ...

```

**Comment by annulen** [comment]
@tomwilliz There is no compilation error in your fragment. Please post actuall error messages. Thanks!

**Comment by tomwilliz** [comment]
what do you mean? is this not an error: "GraphicsContext3DOpenGL.cpp(195): error C2065: 'GL_RGBA8': undeclared identifier"?

**Comment by annulen** [comment]
Sorry, I see now

**Comment by annulen** [comment]
`GraphicsContext3DOpenGL.cpp` should not be compiled in ANGLE-only configuration,  `GraphicsContext3DOpenGLES.cpp` should be used instead. Please use `-DQT_USES_GLES2_ONLY=ON` to activate this code path.

AFAIU it's currently impossible to distinguish pure-ANGLE Qt and Dynamic GL Qt from CMake without crude hacks like invoking qmake with fake project file, or parsing .pri files. In case QtWebKit is built as a qt5 submodule, qmake wrapper will pass QT_USES_GLES2_ONLY automatically.

**Comment by sinspiral** [comment]
@annulen That's good to know!! I was dealing with that on aed012c but I didn't realize about that should be ON by default.

I'm gonna try that with my changesets to see if it would fix it.

**Comment by annulen** [comment]
Sorry, I should probably leave some better comment there :) Originally I wanted to name this variable `QT_USES_ANGLE_ONLY`, but it actually does not control anything ANGLE-specific, but on;y OpenGL ES-specific stuff

**Comment by annulen** [comment]
I.e. it should work exactly the same way in case someone uses other OpenGL ES implementation on Windows than ANGLE.

**Comment by tomwilliz** [comment]
I have set -DQT_USES_GLES2_ONLY=ON, GraphicsContext3DOpenGLES.cpp is compiled now, but I get new errors:

```
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(45): error C2039: 'glReleaseShaderCompiler': is not a member of '`global namespace'' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(45): error C3861: 'glReleaseShaderCompiler': identifier not found [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(56): error C2039: 'glBindFramebuffer': is not a member of '`global namespace'' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(56): error C3861: 'glBindFramebuffer': identifier not found [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(63): error C2039: 'glBindFramebuffer': is not a member of '`global namespace'' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(63): error C3861: 'glBindFramebuffer': identifier not found [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(99): error C2039: 'glBindFramebuffer': is not a member of '`global namespace'' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(99): error C3861: 'glBindFramebuffer': identifier not found [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(105): error C2039: 'glFramebufferTexture2D': is not a member of '`global namespace'' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(105): error C3861: 'glFramebufferTexture2D': identifier not found [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(113): error C2440: 'static_cast': cannot convert from 'WebCore::Extensions3D *' to 'WebCore::Extensions3DOpenGLES *' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(114): error C2027: use of undefined type 'WebCore::Extensions3DOpenGLES' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(114): error C2227: left of '->isImagination' must point to class/struct/union/generic type [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(119): error C2027: use of undefined type 'WebCore::Extensions3DOpenGLES' [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]
  D:\Qt\webkit-qtwebkit-tp5\Source\WebCore\platform\graphics\opengl\GraphicsContext3DOpenGLES.cpp(119): error C2227: left of '->framebufferTexture2DMultisampleIMG' must point to class/struct/union/generic type [D:\Qt\webkit-qtwebkit-tp5\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]

...
```

Any ideas?

**Comment by annulen** [comment]
See opening post, something is wrong with enabling macros like `USE_OPENGL_ES_2`, probably their ordering

**Comment by tomwilliz** [comment]
yes indeed, but even without that, there are still errors.

**Comment by annulen** [comment]
Look how legacy version (e.g. from 5.8 git branch, or 5.8.0 source archive; if you use older Qt, use respective version) is building in your environment. Does it build GraphicsContext3DOpenGLES.cpp and Extensions3DOpenGLES.h or not, what #if's are active and from what headers macros come.

**Comment by annulen** [comment]
Should be fixed after 8fd4477b58e2bf1dda9e8da815c6e57d28e2881d and 7ec8dd4af5828b82e4bc6cdeec4f590b4311153a

**Comment by annulen** [comment]
Update: since 7d5a7eddff62b8db5c9965f7b935b94294b60185 dynamic gl should be autodetected, no need for QT_USES_GLES2_ONLY
